core/deadline.go

```diff
@@ -34,17 +34,106 @@ type slotTimeProvider interface {
 	eth2client.SlotDurationProvider
 }
 
-// Deadliner provides duty Deadline functionality. The C method isn’t thread safe and may only be used by a single goroutine. So, multiple
-// instances are required for different components and use cases.
+// Deadliner provides duty Deadline functionality. The C method isn’t thread safe and
+// may only be used by a single goroutine. So, multiple instances are required
+// for different components and use cases.
 type Deadliner interface {
-	// Add adds a duty to be notified of the Deadline via C. Note that duties will be deduplicated and only a single duty will be provided via C.
+	// Add adds a duty to be notified of the Deadline via C.
+	// Note that duties will be deduplicated and only a single duty will be provided via C.
 	Add(duty Duty)
 
 	// C returns the same read channel every time and contains deadlined duties.
 	// It should only be called by a single goroutine.
 	C() <-chan Duty
 }
 
+// Deadline implements the Deadliner interface.
+type Deadline struct {
+	dutyChan     chan Duty
+	deadlineChan chan Duty
+	quit         chan struct{}
+	deadlineFunc func(Duty) time.Time
+}
+
+// NewDeadliner returns a new instance of Deadline.
+// It runs a goroutine which is responsible for reading and storing duties,
+// and sending the deadlined duty to receiver's deadlineChan.
+func NewDeadliner(ctx context.Context, deadlineFunc func(Duty) time.Time) (*Deadline, error) {
+	d := &Deadline{
+		dutyChan:     make(chan Duty),
+		deadlineChan: make(chan Duty),
+		quit:         make(chan struct{}),
+		deadlineFunc: deadlineFunc,
+	}
+
+	go func() {
+		duties := make(map[Duty]bool)
+		currDuty, currDeadline := getCurrDuty(duties, d.deadlineFunc)
+		currTimer := time.NewTimer(time.Until(currDeadline))
+
+		defer func() {
+			close(d.quit)
+			currTimer.Stop()
+		}()
+
+		setCurrState := func() {
+			if !currDeadline.IsZero() {
+				currTimer.Stop()
+			}
+			currDuty, currDeadline = getCurrDuty(duties, d.deadlineFunc)
+			currTimer = time.NewTimer(time.Until(currDeadline))
+		}
+
```

Comment:
add a TODO here maybe: `optimise getCurrDuty and updating current state if earlier deadline detected`